File: packages/shop/src/services/fiat-order-post-payment-worker.ts
import { processWithConcurrencyLimit } from '@ltdjms/shared';
import type { FiatOrderRepository } from '../domain/fiat-order-repository.js';
import type { FiatOrder } from '../domain/fiat-order.js';
import {
  isBuyerNotified,
  shouldAutoCreateEscortOrder,
  isAdminNotified,
  hasFulfillmentReward,
  isRewardGranted,
  toFulfillmentProduct,
} from '../domain/fiat-order.js';
import type { DispatchOrderSnapshot, EscortDispatchHandoffService } from '../domain/escort-dispatch-handoff-service.js';
import type { EscortOrderBuyerNotifier, AdminOrderNotifier, ProductRewardGranter } from '../domain/notification-interfaces.js';
import pino from 'pino';

/**
 * Exception indicating a workflow state violation that prevents processing.
 * The operation should be released and retried rather than logged as a failure.
 */
export class WorkflowStateException extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'WorkflowStateException';
  }
}

const DEFAULT_BATCH_SIZE = 20;

export class FiatOrderPostPaymentWorker {
  private readonly log: pino.Logger;

  constructor(
    private readonly fiatOrderRepository: FiatOrderRepository,
    private readonly buyerNotificationService: { notifyPaymentSucceeded(order: FiatOrder): void },
    private readonly escortDispatchHandoffService: EscortDispatchHandoffService,
    private readonly escortOrderBuyerNotificationService: EscortOrderBuyerNotifier,
    private readonly adminNotificationService: AdminOrderNotifier,
    private readonly productRewardService: ProductRewardGranter,
    logger?: pino.Logger,
  ) {
    this.log = logger ?? pino({ level: 'warn' });
  }

  async processPendingOrders(): Promise<void> {
    const orders = await this.fiatOrderRepository.findOrdersPendingPostPayment(DEFAULT_BATCH_SIZE);
    await processWithConcurrencyLimit(orders, order => this.processSingleOrder(order), 5);
  }

  async processSingleOrder(order: FiatOrder): Promise<void> {
    const claimTime = new Date();
    const claimed = await this.fiatOrderRepository.claimFulfillmentProcessing(
      order.orderNumber,
      claimTime,
    );
    if (!claimed) return;

    try {
      const fulfillmentProduct = toFulfillmentProduct(order);

      // Step 1: Buyer notification (idempotent)
      if (!isBuyerNotified(order)) {
        this.buyerNotificationService.notifyPaymentSucceeded(order);
        await this.fiatOrderRepository.markBuyerNotifiedIfNeeded(order.orderNumber, new Date());
      }

      // Step 2: Escort handoff (conditional)
      if (shouldAutoCreateEscortOrder(order) && !isAdminNotified(order)) {
        const handoffResult = await this.escortDispatchHandoffService.handoffFromFiatPayment(
          order.guildId,
          order.buyerUserId,
          fulfillmentProduct,
          order.orderNumber,
        );
        if (!handoffResult.isOk()) {
          throw new WorkflowStateException(handoffResult.getError().message);
        }

        const dispatchOrder = handoffResult.getValue();
        const adminClaimTime = new Date();
        const adminClaimed = await this.fiatOrderRepository.claimAdminNotificationProcessing(
          order.orderNumber,
          adminClaimTime,
        );
        if (adminClaimed) {
          try {
            this.escortOrderBuyerNotificationService.notifyEscortOrderCreated(dispatchOrder);
            this.adminNotificationService.notifyAdminsOrderCreated(
              dispatchOrder.guildId,
              dispatchOrder.customerUserId,
              dispatchOrder,
            );
            await this.fiatOrderRepository.markAdminNotifiedIfNeeded(
              order.orderNumber,
              adminClaimTime,
            );
          } catch (e) {
            await this.fiatOrderRepository.releaseAdminNotificationProcessing(order.orderNumber);
            throw e;
          }
        } else {
          this.log.warn(
            { orderNumber: order.orderNumber },
            'Admin notification claim failed, another worker is processing',
          );
        }
      }

      // Step 3: Reward grant (idempotent)
      if (hasFulfillmentReward(order) && !isRewardGranted(order)) {
        const rewardResult = await this.productRewardService.grantReward({
          guildId: order.guildId,
          userId: order.buyerUserId,
          product: fulfillmentProduct,
          amount: fulfillmentProduct.rewardAmount!,
          description: `法幣商品獎勵: ${fulfillmentProduct.name}`,
        });
        if (rewardResult.isErr()) {
          throw new WorkflowStateException(rewardResult.getError().message);
        }
        await this.fiatOrderRepository.markRewardGrantedIfNeeded(
          order.orderNumber,
          new Date(),
        );
      }

      // Step 4: Mark fulfilled (Java: no release on null return)
      await this.fiatOrderRepository.markFulfilledIfNeeded(order.orderNumber, new Date());
    } catch (e) {
      await this.fiatOrderRepository.releaseFulfillmentProcessing(order.orderNumber);
      if (e instanceof WorkflowStateException) {
        this.log.warn({ orderNumber: order.orderNumber, error: e }, 'Workflow state violation processing paid fiat order');
      } else {
        this.log.error({ orderNumber: order.orderNumber, error: e }, 'Failed to process paid fiat order');
      }
    }
  }
}

File: src/main/java/ltdjms/discord/shop/services/FiatOrderPostPaymentWorker.java
package ltdjms.discord.shop.services;

import java.time.Clock;
import java.time.Instant;
import java.util.List;
import java.util.Objects;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import ltdjms.discord.dispatch.domain.EscortDispatchOrder;
import ltdjms.discord.dispatch.services.EscortDispatchHandoffService;
import ltdjms.discord.membership.services.MembershipSpendService;
import ltdjms.discord.product.services.ProductRewardService;
import ltdjms.discord.shared.DomainError;
import ltdjms.discord.shared.Result;
import ltdjms.discord.shop.domain.FiatOrder;
import ltdjms.discord.shop.domain.FiatOrderRepository;

/** Processes paid fiat orders asynchronously and idempotently. */
public class FiatOrderPostPaymentWorker {

  private static final Logger LOG = LoggerFactory.getLogger(FiatOrderPostPaymentWorker.class);
  private static final int DEFAULT_BATCH_SIZE = 20;

  private final FiatOrderRepository fiatOrderRepository;
  private final ProductRewardService productRewardService;
  private final EscortDispatchHandoffService escortDispatchHandoffService;
  private final ShopAdminNotificationService adminNotificationService;
  private final FiatOrderBuyerNotificationService buyerNotificationService;
  private final EscortOrderBuyerNotificationService escortOrderBuyerNotificationService;
  private final MembershipSpendService membershipSpendService;
  private final Clock clock;

  public FiatOrderPostPaymentWorker(
      FiatOrderRepository fiatOrderRepository,
      ProductRewardService productRewardService,
      EscortDispatchHandoffService escortDispatchHandoffService,
      ShopAdminNotificationService adminNotificationService,
      FiatOrderBuyerNotificationService buyerNotificationService,
      EscortOrderBuyerNotificationService escortOrderBuyerNotificationService,
      MembershipSpendService membershipSpendService) {
    this(
        fiatOrderRepository,
        productRewardService,
        escortDispatchHandoffService,
        adminNotificationService,
        buyerNotificationService,
        escortOrderBuyerNotificationService,
        membershipSpendService,
        Clock.systemUTC());
  }

  FiatOrderPostPaymentWorker(
      FiatOrderRepository fiatOrderRepository,
      ProductRewardService productRewardService,
      EscortDispatchHandoffService escortDispatchHandoffService,
      ShopAdminNotificationService adminNotificationService,
      FiatOrderBuyerNotificationService buyerNotificationService,
      EscortOrderBuyerNotificationService escortOrderBuyerNotificationService,
      MembershipSpendService membershipSpendService,
      Clock clock) {
    this.fiatOrderRepository = Objects.requireNonNull(fiatOrderRepository);
    this.productRewardService = Objects.requireNonNull(productRewardService);
    this.escortDispatchHandoffService = Objects.requireNonNull(escortDispatchHandoffService);
    this.adminNotificationService = Objects.requireNonNull(adminNotificationService);
    this.buyerNotificationService = Objects.requireNonNull(buyerNotificationService);
    this.escortOrderBuyerNotificationService =
        Objects.requireNonNull(escortOrderBuyerNotificationService);
    this.membershipSpendService = Objects.requireNonNull(membershipSpendService);
    this.clock = Objects.requireNonNull(clock);
  }

  public void processPendingOrders() {
    List<FiatOrder> orders = fiatOrderRepository.findOrdersPendingPostPayment(DEFAULT_BATCH_SIZE);
    for (FiatOrder order : orders) {
      processSingleOrder(order);
    }
  }

  void processSingleOrder(FiatOrder order) {
    Instant claimTime = Instant.now(clock);
    if (!fiatOrderRepository.claimFulfillmentProcessing(order.orderNumber(), claimTime)) {
      return;
    }

    try {
      var fulfillmentProduct = order.toFulfillmentProduct();

      if (!order.isBuyerNotified()) {
        buyerNotificationService.notifyPaymentSucceeded(order);
        fiatOrderRepository.markBuyerNotifiedIfNeeded(order.orderNumber(), Instant.now(clock));
      }

      if (order.shouldAutoCreateEscortOrder() && !order.isAdminNotified()) {
        Result<EscortDispatchOrder, DomainError> handoffResult =
            escortDispatchHandoffService.handoffFromFiatPayment(
                order.guildId(), order.buyerUserId(), fulfillmentProduct, order.orderNumber());
        if (handoffResult.isErr()) {
          throw new IllegalStateException(handoffResult.getError().message());
        }

        EscortDispatchOrder dispatchOrder = handoffResult.getValue();
        Instant adminClaimTime = Instant.now(clock);
        if (fiatOrderRepository.claimAdminNotificationProcessing(
            order.orderNumber(), adminClaimTime)) {
          try {
            // 買家通知放在 admin claim 成功之後，避免重試時重複通知買家
            escortOrderBuyerNotificationService.notifyEscortOrderCreated(dispatchOrder);
            adminNotificationService.notifyAdminsOrderCreated(
                dispatchOrder.guildId(), dispatchOrder.customerUserId(), dispatchOrder);
            fiatOrderRepository.markAdminNotifiedIfNeeded(order.orderNumber(), adminClaimTime);
          } catch (Exception e) {
            fiatOrderRepository.releaseAdminNotificationProcessing(order.orderNumber());
            throw e;
          }
        } else {
          throw new IllegalStateException(
              "Fiat admin notification is already being processed: orderNumber="
                  + order.orderNumber());
        }
      }

      if (order.hasFulfillmentReward() && !order.isRewardGranted()) {
        Result<ProductRewardService.RewardGrantResult, DomainError> rewardResult =
            productRewardService.grantReward(
                new ProductRewardService.RewardGrantRequest(
                    order.guildId(),
                    order.buyerUserId(),
                    fulfillmentProduct,
                    fulfillmentProduct.rewardAmount(),
                    String.format("法幣商品獎勵: %s", fulfillmentProduct.name()),
                    ltdjms.discord.currency.domain.CurrencyTransaction.Source.PRODUCT_REWARD,
                    ltdjms.discord.gametoken.domain.GameTokenTransaction.Source.PRODUCT_REWARD));
        if (rewardResult.isErr()) {
          throw new IllegalStateException(rewardResult.getError().message());
        }
        fiatOrderRepository.markRewardGrantedIfNeeded(order.orderNumber(), Instant.now(clock));
      }

      membershipSpendService.recordFiatEscortPayment(order, fulfillmentProduct);
      fiatOrderRepository.markFulfilledIfNeeded(order.orderNumber(), Instant.now(clock));
    } catch (Exception e) {
      fiatOrderRepository.releaseFulfillmentProcessing(order.orderNumber());
      LOG.warn("Failed to process paid fiat order: orderNumber={}", order.orderNumber(), e);
    }
  }
}
